tox.ini

```diff
@@ -38,3 +38,43 @@ commands={envpython} -b {toxinidir}/runtests.py --mode=full {posargs:}
 [testenv:debug]
 basepython=python-dbg
 commands=gdb --args {envpython} {toxinidir}/runtests.py --mode=full {posargs:}
+
+
+[flake8]
+### Flake8 README ###
+# If you are coming here because flake8 is flagging something that it should not
+# (and the flagged problem is not reasonable, i.e. cannot be resolved and still
+# accomplish the coding goal), please add "# noqa" to that line in code if it is
+# only a few lines, or add a "per-file-ignore" in this file if there are
+# multiple problems with one specific file.
+#
+# Adding an error code to the "ignore" directive makes it global, and adding a
+# file to "exclude" (not shown at time of writing) means that the file will not
+# be checked for any errors. Both of these should be avoided without good
+# reason, as they immediately reduce the value provided by this checker.
+#
+# E: pycodestyle errors
+# F: flake8 pyflakes
+# N: pep8 naming
+# W: pycodestyle warnings
+# B: bugbear
+# B9: bugbear opinions
+# B950: bugbear line length, better than E501 since it allows long URLs
+# and slight overages where needed
+select = C, E, F, W, B, B9, B950
+# E203: Whitespace before :, black is pep8 and flake8 is wrong here
+# E501: Line too long. Handled by B950 and Black
+# N801: Ignore rule that class names as UpperCamelCase
+# W503: Line break before binary operator
+ignore = E203, E501, N801, W503
+# Creates C901: max-complexity, follows cyclomatic guidelines.
+# May need to be disabled in some files
+# https://en.wikipedia.org/wiki/Cyclomatic_complexity
+max-complexity = 10
+max-line-length = 88
+per-file-ignores =
+    # Ignore unused import error; most are for reexporting.
+    **/__init__.py:F401
+    # import not at top of file; docstring-related
+    runtests.py:E402
+    setup.py:E402
```

Comment:
I am concerned that these linting checks are only used for tox. What happens when running the checks outside tox?

Reply:
`flake8` looks at `tox.ini` for its configuration, whether `tox` is running it or not. It's become a common place for such tools to put their configuration.

Reply:
I agree with Matti on this, I think the concern here is the difference in flake8 configurations. For example, the CI uses `pycodestyle` with explict config file. 
https://github.com/numpy/numpy/blob/0664ec70f7c10c97650bab9e1fbf1b9571e2ef13/tools/linter.py#L61
All though I agree with `tox.ini` as the source of truth, it should be the only source of truth, we can get rid of https://github.com/numpy/numpy/blob/main/tools/lint_diff.ini

Reply:
I don't disagree - but we can't really begin enforcing flake8 directly, the codebase needs some work. The current pycodestlye config disables a ton of things that black will automatically fix, so there isn't much sense in keeping those. But then flake8 adds pyflakes + bugbear, which don't care about style, but do add real value to the code by catching potential bugs, and are currently not checked at all (currently there are about [14k errors](https://github.com/numpy/numpy/pull/21449#issuecomment-1159370412), black will fix about half but that's still a lot).

If we do apply black to everything at some point and enforce it, that eliminates the need for `pycodestyle` and the current config file. My thought was then that full `flake8` could start being enforced on a subset of files where it passes, and gradually go through the other files to make fixes needed to add them to the "enforced" list. Eventually, it should be possible to just check everything. There will be some check overrides needed, but hopefully we could keep those down to the lines where they're needed (as I politely request [here](https://github.com/numpy/numpy/blob/794131fbbef4d3bf464a70ec400da985b9684dfd/tox.ini#L45-L54))

Reply:
Also guess I never answered why `tox.ini` for the config - `.flake8`, `tox.ini`, and `setup.cfg` are the three options (the maintainer is sadly [pretty against](https://github.com/PyCQA/flake8/issues/234) `pyproject.toml`, otherwise that would have been nice). `setup.cfg` is sort of going away, so I just picked to use an existing file.

Reply:
Ah ok got it, that explains why the two configs are different. All though I'm not sure if we'll ever do that commit at some point to fix the 14k errors.